tests/bluetooth/audio/mocks/src/gatt.c

```diff
@@ -0,0 +1,223 @@
+{
+	struct bt_gatt_notify_params *copy;
+
+	for (unsigned int i = 0; i < mock_bt_gatt_notify_cb_fake.call_count; i++) {
+		copy = mock_bt_gatt_notify_cb_fake.arg1_history[i];
+
+		/* Free UUID deep copy */
+		if (copy->uuid) {
+			free((void *)copy->uuid);
+		}
+
+		free(copy);
+	}
+}
+
+void mock_bt_gatt_cleanup(void)
+{
+	notify_params_deep_copy_destroy();
+}
+
+static struct bt_uuid *uuid_deep_copy(const struct bt_uuid *uuid)
+{
+	struct bt_uuid *copy;
+
+	switch (uuid->type) {
+	case BT_UUID_TYPE_16:
+		copy = malloc(sizeof(struct bt_uuid_16));
+		zassert_not_null(copy);
+		memcpy(copy, uuid, sizeof(struct bt_uuid_16));
+		break;
+	case BT_UUID_TYPE_32:
+		copy = malloc(sizeof(struct bt_uuid_32));
+		zassert_not_null(copy);
+		memcpy(copy, uuid, sizeof(struct bt_uuid_32));
+		break;
+	case BT_UUID_TYPE_128:
+		copy = malloc(sizeof(struct bt_uuid_128));
+		zassert_not_null(copy);
+		memcpy(copy, uuid, sizeof(struct bt_uuid_128));
+		break;
+	default:
+		zassert_unreachable("Unexpected uuid->type 0x%02x", uuid->type);
+	}
+
+	return copy;
+}
+
+static struct bt_gatt_notify_params *notify_params_deep_copy(struct bt_gatt_notify_params *params)
+{
+	struct bt_gatt_notify_params *copy;
+
+	copy = malloc(sizeof(*params));
+	zassert_not_null(copy);
+
+	memcpy(copy, params, sizeof(*params));
+
+	if (params->uuid != NULL) {
+		copy->uuid = uuid_deep_copy(params->uuid);
+	}
+
+	return copy;
+}
+
+int bt_gatt_notify_cb(struct bt_conn *conn, struct bt_gatt_notify_params *params)
+{
+	zassert_not_null(params, "'%s()' was called with incorrect '%s' value", __func__, "params");
+
+	/* Either params->uuid, params->attr, or both has to be provided */
+	zassert_true(params->uuid != NULL || params->attr != NULL,
+		     "'%s()' was called with incorrect '%s' value", __func__,
+		     "params->uuid or params->attr");
+
+	return mock_bt_gatt_notify_cb(conn, notify_params_deep_copy(params));
+}
+
+void bt_gatt_notify_cb_reset(void)
+{
+	notify_params_deep_copy_destroy();
+
+	RESET_FAKE(mock_bt_gatt_notify_cb);
+}
+
+#define foreach_attr_type_dyndb(...)
+#define last_static_handle BT_ATT_LAST_ATTRIBUTE_HANDLE
+
+/* Exact copy of subsys/bluetooth/host/gatt.c:gatt_foreach_iter() */
+static uint8_t gatt_foreach_iter(const struct bt_gatt_attr *attr,
+				 uint16_t handle, uint16_t start_handle,
+				 uint16_t end_handle,
+				 const struct bt_uuid *uuid,
+				 const void *attr_data, uint16_t *num_matches,
+				 bt_gatt_attr_func_t func, void *user_data)
+{
+	uint8_t result;
+
+	/* Stop if over the requested range */
+	if (handle > end_handle) {
+		return BT_GATT_ITER_STOP;
+	}
+
+	/* Check if attribute handle is within range */
+	if (handle < start_handle) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	/* Match attribute UUID if set */
+	if (uuid && bt_uuid_cmp(uuid, attr->uuid)) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	/* Match attribute user_data if set */
+	if (attr_data && attr_data != attr->user_data) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	*num_matches -= 1;
```

Comment:
Any chance this is 0 when this function is called?

Reply:
No. Check the line 194

Reply:
Right, so first time this is called, it will always be non-0, but we are decrementing each time it is called, and we only ever check the value again. 

So if the value is `5` the first time, and `gatt_foreach_iter` is called e.g. 6 times where the above check passes, don't we risk this underflowing?

Reply:
At least to me it seems like we need to add a check for `*num_matches == 0` in this.

Reply:
There is already such check few lines below (176)

Reply:
No, the function that calls this cannot be called with `num_matches == 0`, but `num_matches` is decremented multiple times, and there doesn't seem to be anything stopping it from decrementing it from 3 -> 2 -> 1 -> 0 -> Underflow

Reply:
Let's consider num_matches  = 1;
```
        *num_matches -= 1;

        // now *num_matches is 0;

	result = func(attr, handle, user_data);

	if (!*num_matches) {
		return BT_GATT_ITER_STOP;    // <- return
	}
```

Reply:
Ah, I finally see it (might be the indention that causes this to be hard to read), as I had not seen that in case of 

```c
	if (!*num_matches) {
		return BT_GATT_ITER_STOP;
	}
```

that would trigger a `return` in `bt_gatt_foreach_attr_type`. I had read that `return` as a `break`, which would not work, but a `return` does. 

Nevermind my comment then :) (but perhaps we should split this function into a few smaller ones in the host (and here) to reduce the indentation level to increase readability.